compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3129,6 +3168,131 @@ BType getTypeIntersection(BType lhsType, BType rhsType) {
         }
     }
 
+    private BType getIntersectionForErrorTypes(BType lhsType, BType rhsType) {
+
+        BType detailTypeOne = ((BErrorType) lhsType).detailType;
+        BType detailTypeTwo = ((BErrorType) rhsType).detailType;
+
+        if (isSealedRecord(detailTypeOne) || isSealedRecord(detailTypeTwo)) {
+            return symTable.semanticError;
+        }
+
+        SymbolEnv prevEnv = this.env;
+        this.env = symTable.pkgEnvMap.get(lhsType.tsymbol.owner);
```

Comment:
Given that `Types.java` is like a stateless util class I'm not sure if we should maintain the `env` like this.

Raised the same in @KavinduZoysa's PR too - https://github.com/ballerina-platform/ballerina-lang/pull/26144/files#r520295821.

I think we'll have to refactor `getTypeIntersection` to accept `env` too.

Also, regarding the RHS here, IMO getting the `env` of `lhsType.tsymbol.owner` won't quite be the correct approach if `lhsType` is from an imported module, right? Shouldn't it always be the current module?

Reply:
I think you're correct regarding imported modules issue. If we refactor getTypeIntersection it wouldn't be an issue. Would overloading `getTypeIntersection` work? WDYT?

Reply:
With this change, we would be allowing intersections that were not previously allowed, right? e.g., record and map intersections. This would mean that in other places like type narrowing in the type guard also we will be allowing these intersections. So it is possible that we would need the `env` even then.

Mentioned a similar scenario in https://github.com/ballerina-platform/ballerina-lang/pull/26144/files#r521415590.

Unless there are some specific cases where we know we wouldn't need the `env` (and can pass null), I think we should mandate passing the `env`.

Reply:
Cool. Will go through the usages and figure out what I can do. Thanks for your input :)

Reply:
Fixed